ice.c

```diff
@@ -1430,16 +1427,13 @@ void janus_ice_component_free(GHashTable *components, janus_ice_component *compo
 	}
 	if(component->retransmit_buffer != NULL) {
 		janus_rtp_packet *p = NULL;
-		GList *first = g_list_first(component->retransmit_buffer);
-		while(first != NULL) {
-			p = (janus_rtp_packet *)first->data;
-			first->data = NULL;
-			component->retransmit_buffer = g_list_delete_link(component->retransmit_buffer, first);
+		while(g_queue_get_length(component->retransmit_buffer) > 0) {
```

Comment:
Is it not enough to do
while ((p = (janus_rtp_packet *)g_queue_pop_head(component->retransmit_buffer)) != NULL)
instead of recomputing the length of the queue first?

Reply:
I think I derived this from the code we use for hashtables, were if you iterate on an empty table you get an error. Shouldn't be an issue to use the approach you suggest instead.

Anyway, there's no recomputing: the length of the queue is one of the properties of the GQueue object, which means we're accessing a value that already exists, not traversing the whole queue to see how may items are there.

Reply:
Yes you're right, my mistake, there is no length computing. So it is just the cost of a function call, which is not needed, so yes it is better to remove the call.
No other remarks then :-)

Reply:
Ack, I'll try and update this as soon as I can (a bit busy today). If you can provide some feedback on whether or not the patch as it is gives you improvements with no regression (I especially want to make sure we're not breaking anything with this change), that would definitely help speed up the merging progress! Thanks!

Reply:
Weird, I think I had not spotted this first time:
This time, I get a lot of:
GLib-CRITICAL **: g_queue_peek_tail: assertion 'queue != NULL' failed
as soon as packets are forwarded.

Seems that g_queues have never been allocated.

It was a very simple test, with a videoroom-like plugin, one sender, and one receiver.

Reply:
We create those queues as soon as we start receiving packets, but that will never happen for a sendonly stream, so we should add a check for its existence before processing it. Will fix shortly.

Reply:
Fixed (hopefully).

Reply:
OK great, works good now, no more glib critical errors.
Performance is good also, ice_send_thread is not spotted as a bottleneck by perf anymore

Reply:
Cool, thanks for confirming! Unless there is any objection, I'll make the change you suggested on the length thing and merge, then :+1:

Reply:
Sounds good :-) Thanks for developing this!